### PR TITLE
Break the deserialiser apart

### PR DESCRIPTION
**Pull request by XVincentX**
This PR is the first part of a kind of a big effort for #1135

In particular this is breaking apart the deserialisers to make sure they do not use the classes anymore favouring the functions (which is what we do use in Prism in general) and using types whenever is possible to avoid runtime tests instead.

### CODE REVIEW

File: packages/http/src/validator/index.ts [hunk 1]
```diff
@@ -20,20 +20,16 @@ import { validateSecurity } from './validators/security';
...
 import { findOperationResponse } from './utils/spec';
 import { HttpBodyValidator, HttpHeadersValidator, HttpQueryValidator } from './validators';
 import { NonEmptyArray } from 'fp-ts/NonEmptyArray';
 import { HttpPathValidator } from './validators/path';
 
 export const bodyValidator = new HttpBodyValidator('body');
-export const headersValidator = new HttpHeadersValidator(headerDeserializerRegistry, 'header');
-export const queryValidator = new HttpQueryValidator(queryDeserializerRegistry, 'query');
-export const pathValidator = new HttpPathValidator(pathDeserializerRegistry, 'path');
+export const headersValidator = new HttpHeadersValidator(header, 'header');
```
Comment (XVincentX) [inline packages/http/src/validator/index.ts:30]: This is the next guy I'll dismantle.

File: packages/http/src/validator/validators/__tests__/headers.spec.ts [hunk 2]
```diff
@@ -35,24 +34,6 @@ describe('HttpHeadersValidator', () => {
 
       describe('header is present', () => {
         describe('schema is present', () => {
-          describe('deserializer not available', () => {
```
Comment (karol-maciaszek) [inline packages/http/src/validator/validators/__tests__/headers.spec.ts:38]: Why this test is disabled? The behaviour has changed?

File: packages/http/src/validator/validators/body.ts [hunk 3]
```diff
@@ -28,11 +28,11 @@ export function deserializeFormBody(
...
       if (encoding && encoding.style) {
-        const deserializer = body.get(encoding.style);
-        if (deserializer && schema.properties) {
-          const propertySchema = schema.properties[property];
-          deserialized[property] = deserializer.deserialize(property, decodedUriParams, propertySchema as JSONSchema);
-        }
+        const deserializer = body[encoding.style];
+        const propertySchema = schema.properties?.[property];
+
+        if (propertySchema && typeof propertySchema !== 'boolean')
```
Comment (karol-maciaszek) [inline packages/http/src/validator/validators/body.ts:34]: Does this fix the difference between JSONSchema versions? AFAIK one of them can have schema equal to bool false.

**Review by karol-maciaszek — COMMENTED** [review]
Feels way lighter! Added a couple of comments, most important about named exports and some removed tests.

File: packages/http/src/validator/deserializers/style/delimited.ts [hunk 4]
```diff
@@ -1,40 +1,27 @@
...
     if (Array.isArray(value)) {
       // last query param is taken into account
       value = value[value.length - 1];
     }
 
-    return value ? value.split(this.separator) : '';
+    return value ? value.split(separator) : '';
   }
+
+  return function deserialize(name: string, parameters: IHttpNameValues, schema?: JSONSchema, explode?: boolean) {
```
[the remark this answers is not in the document]
Comment (XVincentX) [reply packages/http/src/validator/deserializers/style/delimited.ts]: Absolutely. Done in 4e79cd0

File: packages/http/src/validator/deserializers/style/__tests__/form.spec.ts [hunk 5]
```diff
@@ -1,114 +1,92 @@
...
-        describe('query param is an array', () => {
-          it('returns unmodified value', () => {
-            expect(
-              formStyleDeserializer.deserialize('key', { key: ['val1', 'val2'] }, { type: 'array' }, true)
-            ).toEqual(['val1', 'val2']);
-          });
+  describe('schema type is an array', () => {
+    describe('explode is set', () => {
+      describe('query param is an array', () => {
+        it('returns unmodified value', () => {
```
[the remark this answers is not in the document]
Comment (XVincentX) [reply packages/http/src/validator/deserializers/style/__tests__/form.spec.ts:14]: a63665e

File: packages/http/src/validator/deserializers/style/__tests__/matrix.spec.ts [hunk 6]
```diff
@@ -1,111 +1,81 @@
-import { HttpParamStyles } from '@stoplight/types';
-
-import { MatrixStyleDeserializer } from '../matrix';
+import matrix from '../matrix';
```
[the remark this answers is not in the document]
Comment (XVincentX) [reply packages/http/src/validator/deserializers/style/__tests__/matrix.spec.ts]: 4db588e

File: packages/http/src/validator/validators/__tests__/headers.spec.ts (hunk 2 above)
Comment (XVincentX) [reply packages/http/src/validator/validators/__tests__/headers.spec.ts:38]: We do not have the registry anymore where you can dynamically query for a deserialiser. They are all fixed and the Type System makes sure we're selecting only the one we effectively have.

File: packages/http/src/validator/validators/body.ts (hunk 3 above)
Comment (XVincentX) [reply packages/http/src/validator/validators/body.ts:34]: The type system claims this is the definition of `propertySchema`:


image

with such check, I am excluding it's a boolean so it's at least a JSONSchema (the version is yet to determine)

File: packages/http/src/validator/deserializers/index.ts [hunk 7]
```diff
@@ -1,29 +1,17 @@
...
-export const path = new HttpParamDeserializerRegistry([
-  new SimpleStyleDeserializer(),
-  new LabelStyleDeserializer(),
-  new MatrixStyleDeserializer(),
-]);
+import { deserializeSimple as simple } from './style/simple';
+import { deserializeForm as form } from './style/form';
+import { deserializeDeepObject as deepObject } from './style/deepObject';
+import { deserializeLabel as label } from './style/label';
+import { deserializeMatrixStyle as matrix } from './style/matrix';
```
Comment (karol-maciaszek) [inline packages/http/src/validator/deserializers/index.ts:5]: Lines 1-4 are not postfixed with "Style" while "Matrix" is, is it intended?

**Review by karol-maciaszek — CHANGES_REQUESTED** [review]
Minor inconsistency in naming

File: packages/http/src/validator/validators/body.ts (hunk 3 above)
Comment (karol-maciaszek) [reply packages/http/src/validator/validators/body.ts:34]: 👍

File: packages/http/src/validator/deserializers/index.ts (hunk 7 above)
Comment (XVincentX) [reply packages/http/src/validator/deserializers/index.ts:5]: It was not. Fixed in 8e13d24

File: packages/http/src/validator/validators/__tests__/headers.spec.ts (hunk 2 above)
Comment (karol-maciaszek) [reply packages/http/src/validator/validators/__tests__/headers.spec.ts:38]: There is a possibility that deserializer is not found: https://github.com/stoplightio/prism/pull/1430/files#diff-ec5f393de032d67a4f08a12ecde5456cR43

What I mean is that despite the typing system it is still possible that http-spec will sneak some garbage. Our code will deal correctly with the garbage ^ but we don't have a test for such behavior. Ideally, we should just rely on http-spec doing a good job but then comes reality.

Comment (XVincentX) [reply packages/http/src/validator/validators/__tests__/headers.spec.ts:38]: Ok I'll add the test back, but I'll still try tore move this — if Http Spec is terrible, we'll fix it there.

**Review by karol-maciaszek — APPROVED** [review]
Good stuff!

**Pull request merged**